### ISSUE TITLE
Support VS Code Web

### ISSUE THREAD

**Issue by adamw** [scalameta/metals-feature-requests#225]
Currently there seems to be no support for Scala on github.dev.

When looking up the Metals extension it says: "The 'Scala (Metals)' extension is not available in VS Code Web."

I imagine this might not be possible as it would require running sbt using scala.js to get the project's metadata? Probably that's only the first challenge to overcome :)

But then, maybe Scala.JS is the future (or rather, JavaScript as the most popular and widely accessible VM), so it might be a goal worth pursuing.

**Comment by gabro** [comment]
In case I understand this correctly, codespaces are hosted dev containers, so we shouldn’t need any architectural changes, as long as Java is present Metals should be able to bootstrap and work.

I think we may just add `extensionKind: [“workspace”]` to our package.json and hope for the best 😅

I may give this a shot over the weekend

**Comment by gabro** [comment]
Ok, a couple things more I've found after a quick search:

- just setting `extensionKind` doesn't work (we can test it for the published extension using `remote.extensionKind` to override it)
- it seems that the documentation for supporting the VS Code Web is still being updated, based on this comment https://github.com/microsoft/vscode/issues/130628#issuecomment-897705052
- this example may serve as a reference point https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-web-extension-sample

**Comment by gabro** [comment]
One other thing to notice is that in the example above `LanguageClient` comes from a new `/browser` submodule of `vscode-languageclient`:

https://github.com/microsoft/vscode-extension-samples/blob/3b00e1fd9a6eb292fa0e5f09f14a019553fb7564/lsp-web-extension-sample/client/src/browserClientMain.ts#L9

which is apparently contained in an unreleased 8.0.0 version of `vscode-languageclient`.

So there are likely a few moving parts here. I'm still confident we can do it, but it's probably best best to wait for the official documentation first.

**Comment by gabro** [comment]
Ok, one last thing, confusingly Codespaces and GitHub.dev seem to be slightly different beasts.

- If I go on GitHub and open a Codespace, Metals works correctly.

- If I run VSCode Web from github.dev, Metals won't install.

This is the case for many other popular extensions (see a similar point raised here https://github.com/Dart-Code/Dart-Code/issues/3528, for example).

One guess is that VSCode Web requires [virtual workspaces](https://github.com/microsoft/vscode/wiki/Virtual-Workspaces), in which case it may be a blocker for Metals until the LSP virtual file system proposal lands (and even in that case, it's not clear to me how to make a build tool work, without direct file system access).

We'll see :-)

**Comment by adamw** [comment]
@gabro thanks for looking into it! Looks more promising than I thought :) Makes sense that this is implemented using containers and not JS-everywhere as the extensions are probably very varied, and not everything has a JS backend (yet ;) )

**Comment by gabro** [comment]
Ok, after exploring some more, here's a Twitter thread with my understanding on the situation https://twitter.com/gabro27/status/1426817922790563844?s=21

After reading this https://github.com/microsoft/vscode/wiki/Virtual-Workspaces#language-extensions-and-virtual-workspaces I'm convinced Metals won't ever work in a useful way in VS Code Web, since we don't have access to Java nor a file system, so built tools work run (no code navigation, errors, etc) and neither will the Scala compiler (no completions, hovers, etc).

We *may* be able to extract some bits and use Scala.js for those, but it will be for the small stuff. For example, Scalafmt could work as a separate extension since it can be compiled down to Scala.js. It used to be like that before being integrated into Metals.

That said, for anything serious I would recommend spinning up a Codespace, where Metals already works fine.

**Comment by adamw** [comment]
@gabro Thanks a lot, that's very informative :) So my first intuition that it's JS all the way down wasn't all that bad.

Codespaces is a good fallback, indeed. Though not as nice as pressing `.` :)

By the way - compiling `dotc` using Scala.JS and then running the compiler using JS could be an interesting project ;)

**Comment by olafurpg** [comment]
@adamw the challenge is less about cross-building the compiler to JS and more about getting the build tool (sbt) to run with a virtual workspace.

**Comment by gabro** [comment]
The way I see it:

- it would be a huge effort to even have a POC of scalac/dotc running on Scala.js (I pitched the idea a few years back to Sébastien and he didn't seem too convinced)

- even if that's possible, we would still need to make Sbt / Bloop (and whatever other build system we need to support) to work:
  - in a browser
  - with a virtual file system

For the reasons above, I don't think we'll ever see a version of Metals that works in a browser. It would be a massive undertaking, for a very limited benefit (you can work from a Desktop version of VS Code or use a Codespace, both a very good options where Metals is fully supported).

Bottom line, I'm going to close this feature request, since I don't think it will ever see the light, but I'm adding documentation to clearly explain where Metals works and where it doesn't, and why. https://github.com/scalameta/metals/pull/3046

Thanks @adamw for raising the issue, it made discover lots of interesting things :-)

**Comment by adamw** [comment]
Thanks @gabro for the investigation :) Also learned a bit, though probably not as much as you ;)
